### ISSUE TITLE
The File downloads without any issues but the contents are not complete

### ISSUE THREAD

**Issue by Upanishads** [lorenzodifuccia/safaribooks#174]
Hello Loren,

The download completes without any issues, there is no error in the log file.
But the downloaded file content is not complete.
You can try this book Agile Java™: Crafting Code with Test-Driven Development (0131482394).

All Lessons will be incomplete.

Could you please look into it.

Also in all the downloaded books table of content is never complete.

**Comment by Upanishads** [comment]
Hey Loren,

The issue is gone in the latest version of your code, however for ubuntu we get the below SSL error OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'wrong version number')]
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'decryption failed or bad record mac')]

Which can be resolved by changing the request module to take TLSv1 or higher as default.
Please change it
Thanks

**Comment by Upanishads** [comment]
I think it is coming in multiprocessing when it tries to create more processes, heres the trace
Downloading book CSSs... (4 files)                                                                                                        
Process Process-3:
Process Process-2:
Traceback (most recent call last):
Traceback (most recent call last):
  File "/home/user/anaconda3/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/user/anaconda3/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/user/anaconda3/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user/safaribooks.py", line 827, in _thread_download_css
    response = self.requests_provider(url)
  File "/home/user/anaconda3/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user/safaribooks.py", line 421, in requests_provider
    **kwargs
  File "/home/user/safaribooks.py", line 827, in _thread_download_css
    response = self.requests_provider(url)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/home/user/safaribooks.py", line 421, in requests_provider
    **kwargs
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/home/user/anaconda3/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 1321, in getresponse
    response.begin()
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 257, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "<string>", line 2, in raise_from
  File "/home/user/anaconda3/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/contrib/pyopenssl.py", line 294, in recv_into
    return self.connection.recv_into(*args, **kwargs)
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 1321, in getresponse
    response.begin()
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1822, in recv_into
    self._raise_ssl_error(self._ssl, result)
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 296, in begin
    version, status, reason = self._read_status()
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
    _raise_current_error()
  File "/home/user/anaconda3/lib/python3.7/http/client.py", line 257, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
  File "/home/user/anaconda3/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "/home/user/anaconda3/lib/python3.7/site-packages/urllib3/contrib/pyopenssl.py", line 294, in recv_into
    return self.connection.recv_into(*args, **kwargs)
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1822, in recv_into
    self._raise_ssl_error(self._ssl, result)
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
    _raise_current_error()
  File "/home/user/anaconda3/lib/python3.7/site-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'decryption failed or bad record mac')]
OpenSSL.SSL.Error: [('SSL routines', 'ssl3_get_record', 'wrong version number')]

**Comment by TheSnoozer** [comment]
Could you perhaps try the workaround suggested in https://github.com/lorenzodifuccia/safaribooks/issues/163#issuecomment-568148184?

**Comment by Upanishads** [comment]
I didn't get the workaround suggested there, but I look at the comparison of the code mentioned in the comments there and i tried to modify and execute the code with that changes and it didn't work out.

**Comment by diegorodriguezv** [comment]
Hi @Upanishads.
The updated code should look this:
```
    def collect_css(self):
        self.display.state_status.value = -1

        if "win" in sys.platform:
            # TODO
            for css_url in self.css:
                self._thread_download_css(css_url)

        else:
            for css_url in self.css:
                self._thread_download_css(css_url)

    def collect_images(self):
        if self.display.book_ad_info == 2:
            self.display.info("Some of the book contents were already downloaded.\n"
                              "    If you want to be sure that all the images will be downloaded,\n"
                              "    please delete the output direcotry '" + self.BOOK_PATH +
                              "' and restart the program.")

        self.display.state_status.value = -1

        if "win" in sys.platform:
            # TODO
            for image_url in self.images:
                self._thread_download_images(image_url)

        else:
            for image_url in self.images:
                self._thread_download_images(image_url)

```
Keep in mind python uses indentation (the spaces before lines) as part of the syntax, so lines should be aligned correctly for it to work.

**Comment by diegorodriguezv** [comment]
BTW I was able to download the book you mentioned correctly.

**Comment by TheSnoozer** [comment]
Please get the latest version, this remaining css or image downloading issue should be fixed by d339d9d
